rust/private/rustc.bzl

```diff
@@ -35,302 +35,277 @@ CrateInfo = provider(
-          directory. In this case, the working dir should be '.', and the deps
-          will be symlinked into the .deps dir from the runfiles tree.
+      deps: List[Label]: The deps from ctx.attr.deps.
+      name: str: Name of the current target.
 
     Returns:
-      Returns a struct containing the following fields:
-        transitive_crates:
-        transitive_dylibs:
-        transitive_staticlibs:
-        transitive_libs: All transitive dependencies, not filtered by type.
-        setup_cmd:
-        link_search_flags:
-        link_flags:
+      Returns a DepInfo provider.
     """
     direct_crates = depset()
     transitive_crates = depset()
     transitive_dylibs = depset(order = "topological")  # dylib link flag ordering matters.
     transitive_staticlibs = depset()
     for dep in deps:
-        if hasattr(dep, "crate_info"):
+        if CrateInfo in dep:
             # This dependency is a rust_library
-            direct_crates += [dep.crate_info]
-            transitive_crates += [dep.crate_info]
-            transitive_crates += dep.depinfo.transitive_crates
-            transitive_dylibs += dep.depinfo.transitive_dylibs
-            transitive_staticlibs += dep.depinfo.transitive_staticlibs
+            direct_crates += [dep[CrateInfo]]
+            transitive_crates += [dep[CrateInfo]]
+            transitive_crates += dep[DepInfo].transitive_crates
+            transitive_dylibs += dep[DepInfo].transitive_dylibs
+            transitive_staticlibs += dep[DepInfo].transitive_staticlibs
         elif hasattr(dep, "cc"):
             # This dependency is a cc_library
-            transitive_dylibs += [l for l in dep.cc.libs if l.basename.endswith(toolchain.dylib_ext)]
-            transitive_staticlibs += [l for l in dep.cc.libs if l.basename.endswith(toolchain.staticlib_ext)]
+            dylibs = [l for l in dep.cc.libs if l.basename.endswith(toolchain.dylib_ext)]
+            staticlibs = [l for l in dep.cc.libs if l.basename.endswith(toolchain.staticlib_ext)]
+            transitive_dylibs += dylibs
+            transitive_staticlibs += staticlibs
         else:
-            fail("rust targets can only depend on rust_library or cc_library targets.")
-
-    transitive_libs = depset([c.output for c in transitive_crates]) + transitive_staticlibs + transitive_dylibs
+            fail("rust targets can only depend on rust_library or cc_library targets." + str(dep), "deps")
 
-    # Create symlinks pointing to each transitive lib in deps_dir.
-    deps_dir = working_dir + "/" + name + ".deps"
-    setup_cmd = ["rm -rf " + deps_dir + "; mkdir " + deps_dir + "\n"]
-    for lib in transitive_libs:
-        setup_cmd += [_symlink_dep_cmd(lib, deps_dir, in_runfiles)]
+    crate_list = transitive_crates.to_list()
+    transitive_libs = depset([c.output for c in crate_list]) + transitive_staticlibs + transitive_dylibs
+    indirect_crates = depset([crate for crate in crate_list if crate not in direct_crates])
```

Comment:
transitive_libs is everything, as opposed to just crates.

I added to compensate for removing the setup_deps symlink hermiticity, by not passing unnecessary link directories to rustc. (This could be entirely pointless, I'm not attached to it)